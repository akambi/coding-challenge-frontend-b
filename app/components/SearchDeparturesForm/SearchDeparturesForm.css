.search-params-form {
  padding-top: 10px;
  padding-bottom: 10px;
  display: flex;
  flex: 1;
  justify-content: space-between;
  align-items: center;
  width: 100%;
}

.search-params-form div label {
  display: block;
  text-align: left;
}

.search-params-form > div {
  padding: 5px;
  display: flex;
  flex-direction: column;
  align-self: flex-start;
}

.search-params-form > div.submit {
  align-self: center;
}

.search-params-form div.error {
  color: yellow;
}

@media screen and (max-width:700px) {
  .search-params-form { flex-wrap: wrap;  }
  .search-params-form > div { flex-basis: 50%; }
  .search-params-form > div.submit { flex-basis: 100%; }
}
