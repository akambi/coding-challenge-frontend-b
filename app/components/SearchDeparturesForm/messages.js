/*
 * This contains all the text for the SearchDeparturesForm component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'busbud.components.SearchDeparturesForm';

export default defineMessages({
  origin: {
    id: `${scope}.origin`,
    defaultMessage: 'Origin',
  },
  destination: {
    id: `${scope}.destination`,
    defaultMessage: 'Destination',
  },
  outboundDate: {
    id: `${scope}.outboundDate`,
    defaultMessage: 'Date',
  },
  adult: {
    id: `${scope}.adult`,
    defaultMessage: 'Adults',
  },
  quebec: {
    id: `${scope}.quebec`,
    defaultMessage: 'Québec',
  },
  montreal: {
    id: `${scope}.montreal`,
    defaultMessage: 'Montréal',
  },
  chooseUpcomingDate: {
    id: `${scope}.chooseUpcomingDate`,
    defaultMessage: 'Date must be in the future',
  },
  search: {
    id: `${scope}.search`,
    defaultMessage: 'Search',
  },
});
