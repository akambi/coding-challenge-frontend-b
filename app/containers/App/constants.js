export const QUEBEC = 'quebec';
export const MONTREAL = 'montreal';
export const API_KEY = 'PARTNER_BaASYYHxTxuOINEOMWq5GA';

export const CityToGeoHashMapping = {
  quebec: 'f2m673',
  montreal: 'f25dvk',
};

export const CHANGE_SEARCH_PARAMS = 'busbud/App/CHANGE_SEARCH_PARAMS';

export const LOAD_DEPARTURES = 'busbud/App/LOAD_DEPARTURES';
export const LOAD_DEPARTURES_SUCCESS = 'busbud/App/LOAD_DEPARTURES_SUCCESS';
export const LOAD_DEPARTURES_ERROR = 'busbud/App/LOAD_DEPARTURES_ERROR';
