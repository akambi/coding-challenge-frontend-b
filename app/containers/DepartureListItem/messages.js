/*
 * This contains all the text for the DepartureListItem component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'busbud.components.DepartureListItem';

export default defineMessages({
  quebec: {
    id: `${scope}.quebec`,
    defaultMessage: 'Québec',
  },
  montreal: {
    id: `${scope}.montreal`,
    defaultMessage: 'Montréal',
  },
});
