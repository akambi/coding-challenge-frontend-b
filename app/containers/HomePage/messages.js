/*
 * HomePage Messages
 *
 * This contains all the text for the HomePage component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'busbud.containers.HomePage';

export default defineMessages({
  searchDeparturesHeader: {
    id: `${scope}.searchDeparturesHeader`,
    defaultMessage:
      'Search all departures for a given origin city and a given destination city for a given day',
  },
});
