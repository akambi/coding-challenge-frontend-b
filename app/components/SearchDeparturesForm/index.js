/*
 * SearchDeparturesForm
 */

import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import fr from 'date-fns/locale/fr';
import DatePicker, { registerLocale } from 'react-datepicker';

import { FormattedMessage, injectIntl } from 'react-intl';
import { makeSelectLocale } from 'containers/LanguageProvider/store/selectors';

import PropTypes from 'prop-types';
import messages from './messages';
import Toggle from '../Toggle';
import { MONTREAL, QUEBEC } from '../../containers/App/constants';

import './SearchDeparturesForm.css';
import 'react-datepicker/dist/react-datepicker.css';
import StyledButton from '../Button/StyledButton';

registerLocale('fr', fr);

const originList = [QUEBEC];
const destinationList = [MONTREAL];

const validateForm = errors => {
  let valid = true;
  Object.values(errors).forEach(
    // if we have an error string set valid to false
    val => {
      if (val.length > 0) {
        valid = false;
      }
    },
  );
  return valid;
};

class SearchDeparturesForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      searchParams: { ...props.searchParams },
      errors: {
        outboundDate: '',
      },
    };
  }

  handleAdultChange = event => {
    const { value } = event.target;

    this.setState(prevState => ({
      ...prevState,
      searchParams: {
        ...prevState.searchParams,
        adult: value,
      },
    }));
  };

  handleSubmit = event => {
    if (event !== undefined && event.preventDefault) {
      event.preventDefault();
    }

    if (validateForm(this.state.errors)) {
      this.props.onSubmitForm({ ...this.state.searchParams });
    }
  };

  selectOrigin = event => {
    const { value } = event.target;

    this.setState(prevState => ({
      ...prevState,
      searchParams: {
        ...prevState.searchParams,
        origin: value,
      },
    }));
  };

  selectDestination = event => {
    const { value } = event.target;

    this.setState(prevState => ({
      ...prevState,
      searchParams: {
        ...prevState.searchParams,
        destination: value,
      },
    }));
  };

  selectOutboundDate = date => {
    const { intl } = this.props;

    const errorOutboundDate =
      date > new Date()
        ? ''
        : intl.formatMessage({ ...messages.chooseUpcomingDate });

    this.setState(prevState => ({
      errors: {
        outboundDate: errorOutboundDate,
      },
      searchParams: {
        ...prevState.searchParams,
        outboundDate: date,
      },
    }));
  };

  render() {
    const { errors, searchParams } = this.state;
    const { locale } = this.props;
    const outboundDate = searchParams.outboundDate
      ? searchParams.outboundDate
      : Date.now();

    return (
      <div className="search-params-form-container">
        <form className="search-params-form" onSubmit={this.handleSubmit}>
          <div className="search-origin">
            <label htmlFor="origin">
              <FormattedMessage {...messages.origin} />
            </label>
            <Toggle
              id="origin"
              value={searchParams.origin}
              values={originList}
              messages={messages}
              onToggle={this.selectOrigin}
            />
          </div>
          <div className="search-destination">
            <label htmlFor="destination">
              <FormattedMessage {...messages.destination} />
            </label>
            <Toggle
              id="destination"
              value={searchParams.destination}
              values={destinationList}
              messages={messages}
              onToggle={this.selectDestination}
            />
          </div>
          <div className="search-outbound-date">
            <label htmlFor="outboundDate">
              <FormattedMessage {...messages.outboundDate} />
            </label>
            <DatePicker
              id="outboundDate"
              locale={locale}
              selected={new Date(outboundDate)}
              onChange={this.selectOutboundDate}
            />
            {errors.outboundDate.length > 0 && (
              <div className="error">{errors.outboundDate}</div>
            )}
          </div>
          <div className="search-adult">
            <label htmlFor="adult">
              <FormattedMessage {...messages.adult} />
            </label>
            <input
              id="adult"
              type="number"
              min="1"
              value={searchParams.adult}
              onChange={this.handleAdultChange}
            />
          </div>
          <div className="submit">
            <StyledButton type="submit" className="submit-btn">
              <FormattedMessage {...messages.search} />
            </StyledButton>
          </div>
        </form>
      </div>
    );
  }
}

SearchDeparturesForm.propTypes = {
  searchParams: PropTypes.object,
  onSubmitForm: PropTypes.func,
  locale: PropTypes.string.isRequired,
  intl: PropTypes.any,
};

const mapStateToProps = createStructuredSelector({
  locale: makeSelectLocale(),
});

const withConnect = connect(
  mapStateToProps,
  null,
);

export default compose(withConnect)(injectIntl(SearchDeparturesForm));
