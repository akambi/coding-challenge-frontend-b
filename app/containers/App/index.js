import React from 'react';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import Header from 'components/Header';
import Footer from 'components/Footer';

import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 16px;
  flex-direction: column;
`;

export default function App() {
  return (
    <AppWrapper>
      <Helmet
        titleTemplate="%s - Busbud's bus departures"
        defaultTitle="Busbud's bus departures"
      >
        <meta
          name="description"
          content="A simple application that lists all departures for a given origin city and a given destination city for a given day"
        />
      </Helmet>
      <Header />
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <Footer />
      <GlobalStyle />
    </AppWrapper>
  );
}
