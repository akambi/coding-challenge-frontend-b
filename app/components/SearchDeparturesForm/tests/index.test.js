import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import { browserHistory } from 'react-router-dom';

import SearchDeparturesForm from '../index';
import configureStore from '../../../configureStore';
import messages from '../messages';
import { MONTREAL, QUEBEC } from '../../../containers/App/constants';

describe('<SearchDeparturesForm />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render the SearchDeparturesForm form', () => {
    const searchParams = {
      origin: 'quebec',
      destination: 'montreal',
      outboundDate: '2020-12-13',
      adult: 1,
    };
    const submitSpy = jest.fn();

    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <SearchDeparturesForm
            onSubmitForm={submitSpy}
            searchParams={searchParams}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should submit the form if clicked on search', () => {
    const searchParams = {
      origin: 'quebec',
      destination: 'montreal',
      outboundDate: '2020-12-13',
      adult: 1,
    };
    const submitSpy = jest.fn();

    const { container, getByLabelText } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <SearchDeparturesForm
            onSubmitForm={submitSpy}
            searchParams={searchParams}
          />
        </IntlProvider>
      </Provider>,
    );

    const origin = messages.origin.defaultMessage;
    const destination = messages.destination.defaultMessage;
    const outboundDate = messages.outboundDate.defaultMessage;
    const adult = messages.adult.defaultMessage;
    fireEvent.change(getByLabelText(origin), { target: { value: QUEBEC } });
    fireEvent.change(getByLabelText(destination), {
      target: { value: MONTREAL },
    });

    const tomorrow = new Date(Date.now() + 1 * 24 * 3600 * 1000);
    fireEvent.change(getByLabelText(outboundDate), {
      target: { value: new Date(tomorrow).toISOString().split('T')[0] },
    });
    fireEvent.change(getByLabelText(adult), { target: { value: 2 } });
    fireEvent.click(container.querySelector('button[type="submit"]'));

    expect(submitSpy).toHaveBeenCalledTimes(1);
  });

  it('should not submit the form if date is in the past', () => {
    const searchParams = {
      origin: 'quebec',
      destination: 'montreal',
      outboundDate: '2020-11-13',
      adult: 1,
    };
    const submitSpy = jest.fn();

    const { container, getByLabelText } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <SearchDeparturesForm
            onSubmitForm={submitSpy}
            searchParams={searchParams}
          />
        </IntlProvider>
      </Provider>,
    );

    const origin = messages.origin.defaultMessage;
    const destination = messages.destination.defaultMessage;
    const outboundDate = messages.outboundDate.defaultMessage;
    const adult = messages.adult.defaultMessage;
    fireEvent.change(getByLabelText(origin), { target: { value: QUEBEC } });
    fireEvent.change(getByLabelText(destination), {
      target: { value: MONTREAL },
    });
    const yesterday = new Date(Date.now() - 2 * 24 * 3600 * 1000);
    fireEvent.change(getByLabelText(outboundDate), {
      target: { value: new Date(yesterday).toISOString().split('T')[0] },
    });
    fireEvent.change(getByLabelText(adult), { target: { value: 2 } });
    fireEvent.click(container.querySelector('button[type="submit"]'));

    expect(submitSpy).not.toHaveBeenCalled();
  });
});
