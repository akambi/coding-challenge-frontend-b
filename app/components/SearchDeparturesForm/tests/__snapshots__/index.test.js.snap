// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchDeparturesForm /> should render the SearchDeparturesForm form 1`] = `
<div
  class="search-params-form-container"
>
  <form
    class="search-params-form"
  >
    <div
      class="search-origin"
    >
      <label
        for="origin"
      >
        <span>
          Origin
        </span>
      </label>
      <select
        class="Select-qu3sxg-0 fumbYO"
        id="origin"
      >
        <option
          value="quebec"
        >
          Québec
        </option>
      </select>
    </div>
    <div
      class="search-destination"
    >
      <label
        for="destination"
      >
        <span>
          Destination
        </span>
      </label>
      <select
        class="Select-qu3sxg-0 fumbYO"
        id="destination"
      >
        <option
          value="montreal"
        >
          Montréal
        </option>
      </select>
    </div>
    <div
      class="search-outbound-date"
    >
      <label
        for="outboundDate"
      >
        <span>
          Date
        </span>
      </label>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            id="outboundDate"
            type="text"
            value="12/12/2020"
          />
        </div>
      </div>
    </div>
    <div
      class="search-adult"
    >
      <label
        for="adult"
      >
        <span>
          Adults
        </span>
      </label>
      <input
        id="adult"
        min="1"
        type="number"
        value="1"
      />
    </div>
    <div
      class="submit"
    >
      <button
        class="submit-btn StyledButton-e3glmp-0 iVzJYl"
        type="submit"
      >
        <span>
          Search
        </span>
      </button>
    </div>
  </form>
</div>
`;
